### ISSUE TITLE
Generic rules to change widget properties.

### ISSUE THREAD

**Issue by JPHammonds** [caqtdm/caqtdm#44]
I posted this same think on user forum on glitter but not sure if that is active.  This may be more appropriate place for a feature request.  

I know from previous (a few years ago) exchanges with Anton the intent was to to get caQtDM working and providing features of MEDM, rather than to allow defocus of extras from the likes of CSS. It seems that caQtDM is now mature enough that it may be time to consider some new feature requests. One major difference between CSS and MEDM/caQtDM is the use of visibility calc vs a more feature rich set of rules. With CSS they were reinventing things more. At the time I was working on the adl2boy conversion and put in the request for the visibility calc so that I could deal with that part of the conclusion. At the same time someone (Ralph Lange I think) else was looking for a different "rule". Somehow this quickly evolved into the rules as they now are in CSS. The idea is basic, MEDM and now caQtDM only dealt with visibility property, a rule should be able to affect any property of a widget (visibility, position, color, text, etc.). In addition MEDM/caQtDM only deal with 4 channels, a calc etc to determine visibility. CSS allowed a more extensive list of PVs. This is mostly important that each rule has it's list of PVs to deal with. I have users who are looking for features more like the CSS rules and have asked me to turn this into a more formal request.  

Please consider this a request to add "rules" (calcs that get input from channels)  that allow modification of a selectable widget property.  Given that visibility was handled like in MEDM, this should likely be done as separate functionality.  It may in the end replace the visibility stuff but starting here may cause error in what is there.

**Comment by HelgeBrands** [comment]
Hi John,
possibly I did not understand the problem completly you want to solve with rules. Since a long time we added a lot of extra features to the caCalc widget e.g. RegEx and Python. It is true that there a multiple limitations like number channels, but here you can work with softchannels to overcome this limitation. 
The simple access to all features of an object is from our side not the main goal. The simples reason for this is e.g. the access of the alarm color, they should stay as the are. 
To some feature like position or size the are the methode Signal and Slot from QT. This can be seen in the tests.ui. 
For me it sounds like a little bit, that some users needs a scripting engine inside the display manager and this is against the basic idea of such an viewing tool. But I don't want to inhibit any development from your side to integrate a widget that supports your rules.

**Comment by JPHammonds** [comment]
Helge,

I will look more closely at those features.  I was not aware that these may solve similar issues or that Python is possible.  Will look and get back to you.  Thank you for the reply.

Jo

**Comment by JPHammonds** [comment]
Helge,

One problem is finding what these widgets can do.  I am looking at the web page:

http://epics.web.psi.ch/software/caqtdm/caQtDM_Manual

This does not provide much info on caCalc & nothing on RegExp or python.  Is there a more up to date documentation for this.  Would really just like a mechanism to change different properties as is done for visibility.

John

From: "JPHammonds@anl.gov" <jphammonds@anl.gov>
Date: Thursday, September 13, 2018 at 9:57 AM
To: caqtdm/caqtdm <reply@reply.github.com>, caqtdm/caqtdm <caqtdm@noreply.github.com>
Cc: Author <author@noreply.github.com>, "JPHammonds@anl.gov" <jphammonds@anl.gov>
Subject: Re: [caqtdm/caqtdm] Generic rules to change widget properties. (#44)

Helge,

I will look more closely at those features.  I was not aware that these may solve similar issues or that Python is possible.  Will look and get back to you.  Thank you for the reply.

Jo

**Comment by AntonMezger** [comment]
Hi Join,

 

You could find some information in the presentation I made at Argonne this spring. You may find it at the caQtDM site at PSI. Otherwise, you may take a look at tests.ui

 

Best regards

 

Anton

 

De : John Hammonds <notifications@github.com> 
Envoyé : jeudi 13 septembre 2018 17:12
À : caqtdm/caqtdm <caqtdm@noreply.github.com>
Cc : Subscribed <subscribed@noreply.github.com>
Objet : Re: [caqtdm/caqtdm] Generic rules to change widget properties. (#44)

 

Helge, 

One problem is finding what these widgets can do. I am looking at the web page: 

http://epics.web.psi.ch/software/caqtdm/caQtDM_Manual 

This does not provide much info on caCalc & nothing on RegExp or python. Is there a more up to date documentation for this. Would really just like a mechanism to change different properties as is done for visibility. 

John 

From: "JPHammonds@anl.gov" <jphammonds@anl.gov> 
Date: Thursday, September 13, 2018 at 9:57 AM 
To: caqtdm/caqtdm <reply@reply.github.com>, caqtdm/caqtdm <caqtdm@noreply.github.com> 
Cc: Author <author@noreply.github.com>, "JPHammonds@anl.gov" <jphammonds@anl.gov> 
Subject: Re: [caqtdm/caqtdm] Generic rules to change widget properties. (#44) 

Helge, 

I will look more closely at those features. I was not aware that these may solve similar issues or that Python is possible. Will look and get back to you. Thank you for the reply. 

Jo

**Comment by AntonMezger** [comment]
No,

 

Not that I know off

 

Best regards

 

Anton

 

De : John Hammonds <notifications@github.com> 
Envoyé : jeudi 13 septembre 2018 19:24
À : caqtdm/caqtdm <caqtdm@noreply.github.com>
Cc : Anton Mezger <antonmezger@gmail.com>; Comment <comment@noreply.github.com>
Objet : Re: [caqtdm/caqtdm] Generic rules to change widget properties. (#44)

 

Also, 

Do you know if the talk here was recorded? Sometimes we say much more than we write. 

John

**Comment by JPHammonds** [comment]
Anton/Helge,

I have been looking at the tests.ui to get an idea about how the signals_2 panel works.  I see the calcs that are calculating the geometry (x,y,width, height) which are collected by another caCalc which uses the other four caCalcs (calcVariationX, calcVariationY,calcVariationWidth, calcVariationHeight).  This has a calc property which is set to %QRect, which I imagine the packs the others into a QRect object which somehow is used to move the cagraphics object around the screen.  What I seem to be missing here is how this information is passed to the correct widget.  I run tests.ui and it is working, the widgets move one inside the main box the other inside the smaller scroll area.  I

I made a small example based on this, included here.  Just something to generate position & size values, a calc to gather these into a rectangle, a rectangle and a signal propagator.   What am I missing to actually get the rectangle to move?

John

**Comment by JPHammonds** [comment]
Anton & Helge,

Nevermind the last e-mail I just found out how to switch on editing signals and slots.  My only use of designer has been for caQtDm screens so far.  I have always programmed signals and slots manually in code.

Working on this.  RegExp seems to be explained in the presentation.  It may still take me a bit on that but only because my genereal regexp experience is shallow.  Helge also mentioned that Python is also tied in.  How would I do this?  Is there a way to include the python code in a screen or do I need to start by coding in pyqt & load screens and grab hold of things.
John

From: "JPHammonds@anl.gov" <jphammonds@anl.gov>
Date: Thursday, September 13, 2018 at 2:54 PM
To: caqtdm/caqtdm <reply@reply.github.com>, caqtdm/caqtdm <caqtdm@noreply.github.com>
Cc: Author <author@noreply.github.com>, "JPHammonds@anl.gov" <jphammonds@anl.gov>
Subject: Re: [caqtdm/caqtdm] Generic rules to change widget properties. (#44)

Anton/Helge,

I have been looking at the tests.ui to get an idea about how the signals_2 panel works.  I see the calcs that are calculating the geometry (x,y,width, height) which are collected by another caCalc which uses the other four caCalcs (calcVariationX, calcVariationY,calcVariationWidth, calcVariationHeight).  This has a calc property which is set to %QRect, which I imagine the packs the others into a QRect object which somehow is used to move the cagraphics object around the screen.  What I seem to be missing here is how this information is passed to the correct widget.  I run tests.ui and it is working, the widgets move one inside the main box the other inside the smaller scroll area.  I

I made a small example based on this, included here.  Just something to generate position & size values, a calc to gather these into a rectangle, a rectangle and a signal propagator.   What am I missing to actually get the rectangle to move?

John
